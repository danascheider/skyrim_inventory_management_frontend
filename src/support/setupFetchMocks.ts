import { vi } from 'vitest'
import createFetchMock from 'vitest-fetch-mock'

/*
 *
 * API Mock Setup
 *
 */

const fetchMocker = createFetchMock(vi)

fetchMocker.enableMocks()
