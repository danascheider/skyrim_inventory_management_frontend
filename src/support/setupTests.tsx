import { type ReactElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { render } from '@testing-library/react'
import { LoginContext } from '../contexts/loginContext'
import { User } from 'firebase/auth'
import testProfileImg from './testProfileImg.png'

export const requireLogin = () => { /* noop */ }

export const testUser = {
  uid: 'edna',
  displayName: 'Edna St. Vincent Millay',
  email: 'edna@gmail.com',
  photoURL: testProfileImg,
  emailVerified: true,
  getIdToken: () => new Promise<string>((resolve, _reject) => resolve('xxxxxxx'))
} as User

export const renderWithRouter = (ui: ReactElement) =>
  render(<BrowserRouter>{ui}</BrowserRouter>)

export const renderAuthenticated = (
  ui: ReactElement,
  authLoading = false,
  authError = undefined
) =>
  renderWithRouter(
    <LoginContext.Provider value={{ user: testUser, token: 'xxxxxxx', requireLogin, authLoading, authError }}>
      {ui}
    </LoginContext.Provider>
  )

export const renderUnauthenticated = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider value={{ user: null, token: null, authLoading: false, requireLogin }}>
      {ui}
    </LoginContext.Provider>
  )

export const renderAuthLoading = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider value={{ user: null, token: null, authLoading: true, requireLogin }}>
      {ui}
    </LoginContext.Provider>
  )
