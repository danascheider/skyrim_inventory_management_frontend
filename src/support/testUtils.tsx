import { type ReactElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { JSDOM } from 'jsdom'
import { render } from '@testing-library/react'
import { LoginContext } from '../contexts/loginContext'
import { User } from 'firebase/auth'
import testProfileImg from './testProfileImg.png'

export const BASE_APP_URI = 'http://localhost:5173'

declare global {
  namespace NodeJS {
    interface Global {
      document: Document
      window: Window
    }
  }
}

const setDom = () => {
  const dom = new JSDOM('<!doctype html><html><body></body></html>', {
    url: 'http://localhost:5173',
  })

  global.window = dom.window as unknown as Window & typeof globalThis
  global.document = dom.window.document
}

/*
 *
 * Global Test Data
 *
 */

export const requireLogin = () => {
  /* noop */
}

export const testUser = {
  uid: 'edna',
  displayName: 'Edna St. Vincent Millay',
  email: 'edna@gmail.com',
  photoURL: testProfileImg,
  emailVerified: true,
  getIdtoken: () =>
    new Promise<string>((resolve, _reject) => resolve('xxxxxxx')),
} as unknown as User

/*
 *
 * Test Renderers
 *
 */

const renderWithJSDOM = (ui: ReactElement) => {
  setDom()

  return render(ui)
}

export const renderWithRouter = (ui: ReactElement) =>
  renderWithJSDOM(<BrowserRouter>{ui}</BrowserRouter>)

export const renderAuthenticated = (ui: ReactElement, authLoading = false) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{
        user: testUser,
        token: 'xxxxxxx',
        requireLogin,
        authLoading,
      }}
    >
      {ui}
    </LoginContext.Provider>
  )

export const renderUnauthenticated = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{
        user: null,
        token: null,
        authLoading: false,
        requireLogin,
      }}
    >
      {ui}
    </LoginContext.Provider>
  )

export const renderAuthLoading = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{ user: null, token: null, authLoading: true, requireLogin }}
    >
      {ui}
    </LoginContext.Provider>
  )
