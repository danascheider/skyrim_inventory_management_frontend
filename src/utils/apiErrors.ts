export class UnprocessableEntityError extends Error {
  readonly code: 422
  readonly message: string
  readonly name: 'UnprocessableEntityError'

  constructor(message: string = '422 Unprocessable Entity') {
    super(message)

    this.code = 422
    this.name = 'UnprocessableEntityError'
    this.message = message

    Object.setPrototypeOf(this, UnprocessableEntityError.prototype)
  }
}

export class InternalServerError extends Error {
  readonly code: 500
  readonly message: string
  readonly name: 'InternalServerError'

  constructor(message: string = '500 Internal Server Error') {
    super(message)

    this.code = 500
    this.name = 'InternalServerError'
    this.message = message

    Object.setPrototypeOf(this, InternalServerError.prototype)
  }
}
