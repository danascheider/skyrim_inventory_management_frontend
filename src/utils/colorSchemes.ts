export interface ColorScheme {
  schemeColorDarkest: string
  schemeColorDark: string
  schemeColorLight: string
  schemeColorLightest: string
  hoverColorDark: string
  hoverColorLight: string
  hoverColorLightest: string
  textColorPrimary: string
  textColorSecondary: string
  textColorTertiary: string
  borderColor: string
}

export const YELLOW: ColorScheme = {
  schemeColorDarkest: '#ffbf00',
  schemeColorDark: '#ffcb32',
  schemeColorLight: '#ffd866',
  schemeColorLightest: '#fff2cc',
  hoverColorDark: '#e5ab00',
  hoverColorLight: '#ffc519',
  hoverColorLightest: '#ffebb2',
  textColorPrimary: '#4c3900',
  textColorSecondary: '#4c3900',
  textColorTertiary: '#7f5700',
  borderColor: '#cc9800',
}

export const PINK: ColorScheme = {
  schemeColorDarkest: '#e83f6f',
  schemeColorDark: '#ec658b',
  schemeColorLight: '#f18ba8',
  schemeColorLightest: '#fad8e2',
  hoverColorDark: '#d03863',
  hoverColorLight: '#ea527d',
  hoverColorLightest: '#f5b2c5',
  textColorPrimary: '#fff',
  textColorSecondary: '#fff',
  textColorTertiary: '#451221',
  borderColor: '#b93258',
}

export const BLUE: ColorScheme = {
  schemeColorDarkest: '#2274a5',
  schemeColorDark: '#4e8fb7',
  schemeColorLight: '#7aabc9',
  schemeColorLightest: '#d2e3ed',
  hoverColorDark: '#1e6894',
  hoverColorLight: '#3881ae',
  hoverColorLightest: '#bcd5e4',
  textColorPrimary: '#fff',
  textColorSecondary: '#fff',
  textColorTertiary: '#0d2e42',
  borderColor: '#1b5c84',
}

export const GREEN: ColorScheme = {
  schemeColorDarkest: '#00a323',
  schemeColorDark: '#32b54e',
  schemeColorLight: '#66c77b',
  schemeColorLightest: '#ccecd3',
  hoverColorDark: '#00921f',
  hoverColorLight: '#19ac38',
  hoverColorLightest: '#b2e3bd',
  textColorPrimary: '#fff',
  textColorSecondary: '#fff',
  textColorTertiary: '#00410e',
  borderColor: '#00821c',
}

export const AQUA: ColorScheme = {
  schemeColorDarkest: '#20e2e9',
  schemeColorDark: '#62eaef',
  schemeColorLight: '#8ff0f4',
  schemeColorLightest: '#d2f9fa',
  hoverColorDark: '#11cbd1',
  hoverColorLight: '#4ce7ed',
  hoverColorLightest: '#bcf6f8',
  textColorPrimary: '#094345',
  textColorSecondary: '#094345',
  textColorTertiary: '#094345',
  borderColor: '#19b4ba',
}

const colorSchemes: ColorScheme[] = [YELLOW, PINK, BLUE, GREEN, AQUA]

export default colorSchemes
