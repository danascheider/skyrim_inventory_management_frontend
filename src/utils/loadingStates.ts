export type LoadingState = 'LOADING' | 'DONE' | 'ERROR'

export const LOADING: LoadingState = 'LOADING'
export const DONE: LoadingState = 'DONE'
export const ERROR: LoadingState = 'ERROR'
