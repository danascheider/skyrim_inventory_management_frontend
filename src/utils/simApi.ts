import { type Game } from '../types/games'
import { AuthorizationError, InternalServerError, NotFoundError } from './apiErrors'

const baseUri = import.meta.env.PROD ? 'https://sim-api.danascheider.com' : 'http://localhost:5173'

const contentTypeHeader = { 'Content-Type': 'application/json' }
const authHeader = (token: string) => ({ 'Authorization': `Bearer ${token}` })
const combinedHeaders = (token: string) => ({ ...contentTypeHeader, ...authHeader(token) })

export const getGames = (token: string) => {
  const uri = `${baseUri}/games`
  const headers = combinedHeaders(token)

  return (
    fetch(uri, { headers })
      .then((res: Response) => {
        if (res.status === 401) throw new AuthorizationError()
        if (res.status === 404) throw new NotFoundError()
        if (res.status === 500) throw new InternalServerError()
        return res.json().then((json: Game[]) => ({ status: res.status, json }))
      })
  )
}