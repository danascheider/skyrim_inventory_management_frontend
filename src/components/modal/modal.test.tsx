import { describe, expect, test } from 'vitest'
import { renderWithRouter } from '../../support/testUtils'
import { YELLOW } from '../../utils/colorSchemes'
import GameEditForm from '../gameEditForm/gameEditForm'
import Modal from './modal'

describe('Modal', () => {
  test('displays the child', () => {
    const wrapper = renderWithRouter(
      <Modal hidden={false}>
        <GameEditForm
          gameId={4}
          name="My Game"
          description={null}
          buttonColor={YELLOW}
        />
      </Modal>
    )

    expect(wrapper).toBeTruthy()
    expect(wrapper.getByTestId('editGame4')).toBeTruthy()
  })

  test('matches snapshot', () => {
    const wrapper = renderWithRouter(
      <Modal hidden={false}>
        <GameEditForm
          gameId={4}
          name="My Game"
          description={null}
          buttonColor={YELLOW}
        />
      </Modal>
    )

    expect(wrapper).toMatchSnapshot()
  })
})
