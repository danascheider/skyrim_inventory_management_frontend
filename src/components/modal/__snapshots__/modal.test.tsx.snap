// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_root_753dd0"
        data-testid="modal"
      >
        <div
          class="_content_753dd0"
        >
          <div
            class="_root_bc52a1"
          >
            <h3>
              Update Game
            </h3>
            <form
              data-testid="editGame4Form"
              style="--button-background-color: #ffbf00; --button-text-color: #4c3900; --button-hover-color: #e5ab00; --button-border-color: #cc9800;"
            >
              <fieldset
                class="_fieldset_bc52a1"
              >
                <label
                  class="_label_bc52a1"
                  for="name"
                >
                  Name
                </label>
                <input
                  aria-label="Name"
                  class="_input_bc52a1"
                  data-testid="editNameField"
                  name="name"
                  pattern="^\\\\s*[A-Za-z0-9 \\\\-',]*\\\\s*$"
                  title="Name can contain only alphanumeric characters, spaces, commas, hyphens, and apostrophes"
                  type="text"
                  value="My Game"
                />
              </fieldset>
              <fieldset
                class="_fieldset_bc52a1"
              >
                <label
                  class="_label_bc52a1"
                  for="description"
                >
                  Description
                </label>
                <input
                  aria-label="Description"
                  class="_input_bc52a1"
                  data-testid="editDescriptionField"
                  name="description"
                  placeholder="Description"
                  value=""
                />
              </fieldset>
              <button
                class="_submit_bc52a1"
                data-testid="submitGameEditForm"
                type="submit"
              >
                Update
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_root_753dd0"
      data-testid="modal"
    >
      <div
        class="_content_753dd0"
      >
        <div
          class="_root_bc52a1"
        >
          <h3>
            Update Game
          </h3>
          <form
            data-testid="editGame4Form"
            style="--button-background-color: #ffbf00; --button-text-color: #4c3900; --button-hover-color: #e5ab00; --button-border-color: #cc9800;"
          >
            <fieldset
              class="_fieldset_bc52a1"
            >
              <label
                class="_label_bc52a1"
                for="name"
              >
                Name
              </label>
              <input
                aria-label="Name"
                class="_input_bc52a1"
                data-testid="editNameField"
                name="name"
                pattern="^\\\\s*[A-Za-z0-9 \\\\-',]*\\\\s*$"
                title="Name can contain only alphanumeric characters, spaces, commas, hyphens, and apostrophes"
                type="text"
                value="My Game"
              />
            </fieldset>
            <fieldset
              class="_fieldset_bc52a1"
            >
              <label
                class="_label_bc52a1"
                for="description"
              >
                Description
              </label>
              <input
                aria-label="Description"
                class="_input_bc52a1"
                data-testid="editDescriptionField"
                name="description"
                placeholder="Description"
                value=""
              />
            </fieldset>
            <button
              class="_submit_bc52a1"
              data-testid="submitGameEditForm"
              type="submit"
            >
              Update
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
